### ISSUE TITLE
callHttp does no polling

### ISSUE THREAD

**Issue by itpropro** [Azure/azure-functions-durable-js#284]
**Describe the bug**
I have two durable function runtimes running locally on different ports. I try to make a durable http call with `yield context.df.callHttp`.
Regarding to the docs, the orchestrator function should poll based on the return as long as 202 and a location header is returned. 
In my case, callhttp does no polling at all and returns the plain result of the request containing status code 202 and a location header:

```
[2021-08-29T20:17:24.596Z] {
[2021-08-29T20:17:24.597Z]   statusCode: 202,
[2021-08-29T20:17:24.597Z]   headers: {
[2021-08-29T20:17:24.598Z]     Date: 'Sun, 29 Aug 2021 20:17:24 GMT',
[2021-08-29T20:17:24.598Z]     Server: 'Kestrel',
[2021-08-29T20:17:24.599Z]     'Transfer-Encoding': 'chunked',
[2021-08-29T20:17:24.599Z] Executed 'Functions.UpdateFunction' (Succeeded, Id=85f4cbae-34da-4ae8-9f27-469041954a81, Duration=6ms)
[2021-08-29T20:17:24.599Z]     Location: 'http://localhost:7072/runtime/webhooks/durabletask/instances/53fccd02ea564e088b5b03c06520aaf3?taskHub=kametlicensingapihub&connection=DfStorageConnectionString&code=jD7lnp5vqS34wvyYTmSKUpKCWxZy5BXmuCLMPAYCsqGXgAYNUvdhDQ==',
[2021-08-29T20:17:24.600Z]     'Retry-After': '10'
[2021-08-29T20:17:24.601Z]   },
[2021-08-29T20:17:24.601Z]   content: '{\r\n' +
[2021-08-29T20:17:24.602Z]     '  "id": "53fccd02ea564e088b5b03c06520aaf3",\r\n' +
[2021-08-29T20:17:24.603Z]     '  "statusQueryGetUri": "http://localhost:7072/runtime/webhooks/durabletask/instances/53fccd02ea564e088b5b03c06520aaf3?taskHub=kametlicensingapihub&connection=DfStorageConnectionString&code=jD7lnp5vqS34wvyYTmSKUpKCWxZy5BXmuCLMPAYCsqGXgAYNUvdhDQ==",\r\n' +
[2021-08-29T20:17:24.603Z]     '  "sendEventPostUri": "http://localhost:7072/runtime/webhooks/durabletask/instances/53fccd02ea564e088b5b03c06520aaf3/raiseEvent/{eventName}?taskHub=kametlicensingapihub&connection=DfStorageConnectionString&code=jD7lnp5vqS34wvyYTmSKUpKCWxZy5BXmuCLMPAYCsqGXgAYNUvdhDQ==",\r\n' +
[2021-08-29T20:17:24.604Z]     '  "terminatePostUri": "http://localhost:7072/runtime/webhooks/durabletask/instances/53fccd02ea564e088b5b03c06520aaf3/terminate?reason={text}&taskHub=kametlicensingapihub&connection=DfStorageConnectionString&code=jD7lnp5vqS34wvyYTmSKUpKCWxZy5BXmuCLMPAYCsqGXgAYNUvdhDQ==",\r\n' +
[2021-08-29T20:17:24.604Z]     '  "rewindPostUri": "http://localhost:7072/runtime/webhooks/durabletask/instances/53fccd02ea564e088b5b03c06520aaf3/rewind?reason={text}&taskHub=kametlicensingapihub&connection=DfStorageConnectionString&code=jD7lnp5vqS34wvyYTmSKUpKCWxZy5BXmuCLMPAYCsqGXgAYNUvdhDQ==",\r\n' +
[2021-08-29T20:17:24.605Z]     '  "purgeHistoryDeleteUri": "http://localhost:7072/runtime/webhooks/durabletask/instances/53fccd02ea564e088b5b03c06520aaf3?taskHub=kametlicensingapihub&connection=DfStorageConnectionString&code=jD7lnp5vqS34wvyYTmSKUpKCWxZy5BXmuCLMPAYCsqGXgAYNUvdhDQ==",\r\n' +
[2021-08-29T20:17:24.605Z]     '  "restartPostUri": "http://localhost:7072/runtime/webhooks/durabletask/instances/53fccd02ea564e088b5b03c06520aaf3/restart?taskHub=kametlicensingapihub&connection=DfStorageConnectionString&code=jD7lnp5vqS34wvyYTmSKUpKCWxZy5BXmuCLMPAYCsqGXgAYNUvdhDQ=="\r\n' +
[2021-08-29T20:17:24.606Z]     '}'
[2021-08-29T20:17:24.607Z] }
```

I tried calling it with POST and with GET, which made no difference at all:

```
  const result = yield context.df.callHttp(
    "POST",
    `http://localhost:7072/api/OtherFunctionOrchestrator`,
    { a, b, c}
  );
```

```
  const result = yield context.df.callHttp(
    "GET",
    `http://localhost:7072/api/OtherFunctionOrchestrator?parametera=${a}&parameterb=${b}&parameterc=${c}`
  );
```

**Investigative information**

- Durable Functions extension version: ExtensionBundle "[2.*, 3.0.0)"
- durable-functions npm module version: 1.5.2
- Language (JavaScript/TypeScript) and version: TypeScript 4.4.2
- Node.js version: 14.17.4

**Expected behavior**
callHttp should poll the endpoint until the return code is 200 and then return the actual result.

**Actual behavior**
callHttp does one single call, does no polling despite the location header and status 202 being present and returns the first reply package containing the statusUris.

I am using Windows 10 19043.

**Comment by itpropro** [comment]
I thought it may be connected to the extension bundle, but even after removing the bundle and installing all extensions manually with `func extensions install`, callHttp still doesn't poll and returns the initial replay from the other orchestrator.

**Comment by cgillum** [comment]
Pinging @bachuv and @davidmrdavid.

**Comment by davidmrdavid** [comment]
Hi @itpropro,

Thank you for reaching out! Just to make sure we're looking at the same problem, could you please provide a minimal reproducer for this issue? An attached `.zip` or a link to a repro would suffice, and it would help accelerate our investigation; thanks!

**Comment by itpropro** [comment]
Sorry I couldn't get back earlier. I just created a minimal reproduction mono repo with two Azure Functions. They both have local.settings.template.json files that contain my configuration (just adjusted the default port). I did the following to reproduce the error:
1. Started azurite
2. Executed `yarn build && func start` in both folders
3. Executed `Invoke-RestMethod -Method Get -Uri http://localhost:7071/api/Orchestrator1` with pwsh

https://github.com/itpropro/durable-functions-call-http

**Comment by itpropro** [comment]
Any updates @cgillum and @davidmrdavid? Could you reproduce the issue with the repo or do you need anything else from me?

**Comment by davidmrdavid** [comment]
Hi @itpropro! I looked into this today. I can reproduce the issue and I'm currently investigating why this isn't working out of the box. I have some theories already, but nothing I'm ready to share just yet. I'll keep this thread updated, thanks!

**Comment by davidmrdavid** [comment]
Good news! I found the root cause, @itpropro. 

There are two different blockers preventing this behavior at the moment.

The first is that the underlying durable extension (written in C#) requires that durable-http-requests that support polling be marked with a particular boolean flag that indicates the client supports the polling behavior. The JS/TS SDK currently does not send that flag, but it could.

The second is more complicated. Even after setting the right flag in our durable-http-request, the polling behavior introduces a variable number of *history events* in our History array, which we use for orchestration replay and fault tolerance. You can imagine that, in a non-polling scenario, the JS/TS SDK **assumes** that it can retrieve the requests' result from the next http-result event in the History array, but with a polling scenario we make several requests so now there's **extra resolution logic** to be made on the SDK's side to find the event containing the actual result. This **extra task resolution logic** is exactly what we're missing on the SDK's side to enable this feature.

So that's what we need to do to resolve this issue :) 

_However,_ we're currently days away from starting an effort to revamp the JS/TS SDK with the changes described in [this](https://techcommunity.microsoft.com/t5/apps-on-azure/durable-functions-for-python-just-got-faster/ba-p/2647246) post, so I would like to wait until those changes are completed before tackling this feature. These changes will also make it easier to correctly implement this feature's logic, so I think they're a necessary precondition :)

So I'm going to tag this issue as a top-priority after our revamping of the SDK, and update our documentation to indicate this temporary limitation. Is that ok with you?

**Comment by itpropro** [comment]
Thanks for your investigation @davidmrdavid! It's good to hear that this is getting priority in the revamping process, do you already have at least a raw timelime for the revamp process?
Also I am not really sure how to address that in our current project, as we already have multiple Azure Functions based microservices with durable-functions. The only workaround that comes to my mind would be to create a separate activity function like "invokeDurableFunctionRequest" or something similar and use callActivity from the orchestrator. Would that be a viable option from your perspective?

**Comment by davidmrdavid** [comment]
Hi @itpropro,

I think you're right. I believe that, at the moment, the best to go about this would be to manually perform the polling in an Activity instead of relying on the built-in durable http-request. That should allow you to re-use the Activity across orchestrations as well, which is desirable. 

The logic should be relatively simple, essentially just time-limited loop where the request is repeated every x-many seconds and until a non 202 response is received.

As for the timeline, I **expect** to be done with the revamping of the SDK by the second half of September. So I think early October should be a good estimate for when this feature should be enabled. Just a heads up: I might have to guide you through manually installing the latest durable-extension release to truly make this feature available to you as soon as possible, but that's relatively simple :) . 

_EDIT:_ I just saw your edit, so give me a minute to reply to that new message :)

**Comment by davidmrdavid** [comment]
Responding to your edit:

If you expect the polling to take more than 5 minutes, you can implement this within the orchestrator itself by using a combination of a while-loop, the durable http-request API, and timer-triggers. This is essentially what the durable-extension is doing behind the scenes: make a durable http-request inside a while-loop, inspect the result and if the result is a 202, then schedule a timer and go back to the beginning of the loop once a timer fires.

That should allow you to bypass any activity time limitations, and is ultimately closer to what the polling API will do for you behind the scenes one we implement it!

**Comment by davidmrdavid** [comment]
cc/ @ConnorMcMahon, just FYI about this thread.

**Comment by davidmrdavid** [comment]
Just throwing another options here: for modularity, you could implement the polling pattern in a sub-orchestration instead of having it inlined with your pre-existing orchestrators: that would give you more modularity and re-usability :)

**Comment by itpropro** [comment]
Thanks for your feedback, I already have a utility function with caching ready for calling the durable functions, but I think I will go with the sub-orchestrator idea, as the Orchestrators can take longer than 10 minutes under some circumstances. 
I am definitely available and willing to test early versions and provide feedback if needed (NDA are in place anyway ;)).

**Comment by davidmrdavid** [comment]
Sounds good, thank you for your patience and flexibility here. It's my top priority to get this revamping out the door as soon as possible so I'm also trying to get to this feature _asap_. I'll try to keep this thread updated but please reach out if you need anything!

**Comment by davidmrdavid** [comment]
The doc fix was just been merged!
As a note to myself, just noting down that we should **_revert_** this doc change once this feature is merged

**Comment by itpropro** [comment]
Just for the record, this is how I implemented the sub orchestrator and the activity function currently. Keep in mind that it's just a minimal implementation and for me the priority was to be compatible with the callHttp syntax. Would like to hear your thoughts @davidmrdavid :)

Orchestrator (called with `yield context.df.callSubOrchestrator('get', statusQueryGetUri)`)

```Typescript
import * as df from "durable-functions";
import dayjs from "dayjs";

const orchestrator = df.orchestrator(function* (context) {
  const { method, url } = context.df.getInput();
  const pollingInterval = 30;
  const currentTime = dayjs(context.df.currentUtcDateTime);
  const expiryTime = currentTime.add(2, "hour");
  let result = yield context.df.callActivity("MakeHttpRequest", {
    method,
    url,
  });
  while (result.statusCode === 202 && currentTime.isBefore(expiryTime)) {
    result = yield context.df.callActivity("MakeHttpRequest", { method, url });
    const nextCheck = currentTime.add(pollingInterval, "s");
    yield context.df.createTimer(nextCheck.toDate());
  }
  return result.content;
});

export default orchestrator;
```

Activity function (don't use callHttp in the sub orchestrator, as that leads to all kind of strange behavior)

```Typescript
import { AzureFunction, Context } from "@azure/functions";
import axios, { AxiosInstance } from "axios";
// Create a global variable to make it reusable between function calls
let axiosClient: AxiosInstance;
// Configure Azure Functions socket limit (only apply for consumption plan)
import http from "http";
import https from "https";
http.globalAgent.maxSockets = 600;
https.globalAgent.maxSockets = 600;

const activityFunction: AzureFunction = async function (
  context: Context
): Promise<Record<string, unknown>> {
  let { method, url } = context.bindings.input;
  method = method.toLowerCase();
  if (!axiosClient) {
    axiosClient = axios.create();
  }
  let result = null;
  if (method === "get") {
    result = await axiosClient.get(url);
  } else if (method === "post") {
    result = await axiosClient.post(url);
  }
  // Result to be compatible with callHttp result object
  return { statusCode: result?.status, content: result?.data };
};

export default activityFunction;
```

**Comment by davidmrdavid** [comment]
Hi @itpropro,

Thanks for the follow-up :) 
Your suggested implementation makes sense to me, although I'm surprised to hear that `callHttp` leads to "all kind of strange behavior" in the sub-orchestrator. Can you please clarify what you're seeing when you do that?

**Comment by itpropro** [comment]
Of course, as soon as I replace the callActivity with callHttp it produces some weird call loops until the function exits out with “non-deterministic response” error. I tried to reproduce it with node-fetch, axios and native https, but it only appears when using the callHttp method.

**Comment by davidmrdavid** [comment]
That's quite bizarre. I'll keep a note here to investigate further. Thanks!

**Comment by itpropro** [comment]
Hey @davidmrdavid I just wanted to check in, as we are going to continue working on the original project in January. Thanks for the update and happy holidays!

**Comment by davidmrdavid** [comment]
Hi @itpropro,

Thanks for reaching out again! My refactoring work took longer than expected, but it is _functionality complete_ as of a few days ago. Now that I'm back from a small end-of-year break, I'll be spending the next few days just documenting it, and then making a PR for release.

After that, adding the polling option should be fairl quick. Once my refactoring PR is cleaned and validated, I can also expedite a private package for you with the callHttp polling option. 

Let me circle back on progress around mid-week, I've kept this issue in my mind while working on this PR, so I haven't forgotten about it :) . Thanks!

**Comment by davidmrdavid** [comment]
Hi @itpropro, I'm happy to report that I finally opened this major refactoring that I've been teasing for quite a while now - https://github.com/Azure/azure-functions-durable-js/pull/305

After merging it, I should be unblocked to work on this. I'll be back with more updates as we make progress on that front. Thanks!

**Comment by itpropro** [comment]
Hey @davidmrdavid, just checking in, something I can provide feedback on already?

**Comment by davidmrdavid** [comment]
Hi @itpropro,

Sorry for my delayed response, I saw this notification right around the time I got out for a small vacation.
I'm sorry to say that there isn't an implementation yet, there's been a few blockers on various directions.

For starters, I haven't been able to release our new major version of the SDK due to [this](https://github.com/Azure/azure-functions-durable-python/issues/340) blocker still being out. Unfortunately, that's not something us, as the Durable Functions team, have direct influence over; although at this point I'm raising the pressure as much as possible to get it resolved. In case this blocker is confusing: essentially we need a new release of the Durable Functions Extension to be widely available _by default_, and for various reasons external to us that release hasn't been able to proceed. Until this release is out, I can't really prioritize this feature.

@itpropro: just to make sure I understand, you do have a workaround, although an imperfect one, for polling, correct?

**Comment by itpropro** [comment]
Hey @davidmrdavid,
Yes we currently still have the workaround in place we discussed in September with the suborchestrator. It's not optimal and the number of customer projects where it would help to have a native way for function to function communication has grown since.
So if there is anything I can do to support the process of bringing this back, I am happy to help.

**Comment by hossam-nasr** [comment]
Hi @itpropro,

I'm happy to report that this feature has been completed and merged with this PR: #346 ! A new release of the JS SDK is expected very soon, so you should be able to use polling in your code without workarounds soon! Closing this issue for now, but please feel free to reactivate or file a new issue if you continue to face trouble with this after the new SDK release. Thank you for your patience :)
